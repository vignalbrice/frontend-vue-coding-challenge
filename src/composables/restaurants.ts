import { api } from '~/services/api';

export interface Restaurant {
  name: string
  id: string
  url: string
  phone: string
  display_phone: string
  photos: string[]
  location?: Location
  reviews: Review[]
}

export interface Location {
  address1: string
  city: string
  postal_code: string
  formatted_address: string
}

export interface Review {
  text: string
  rating: number
  id: string
}

export function useFetchRestaurants() {
  return useQuery({
    queryKey: [`restaurants-list`],
    queryFn: () => {
      const url = `restaurants`;
      return api(url).json<Restaurant[]>();
    },
  });
}

export function useFetchRestaurant({ restaurantId }: { restaurantId: string | string[] }) {
  return useQuery({
    queryKey: [`company`, restaurantId],
    queryFn: () => {
      const url = `restaurants/${restaurantId}`;
      return api(url).json<Restaurant>();
    },
  });
}
