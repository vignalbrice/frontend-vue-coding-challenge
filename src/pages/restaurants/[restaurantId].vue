<script setup lang="ts">
import { useFetchRestaurant } from '~/composables/restaurants';
import getRestaurantRating from '~/utils/getRestaurantRating';

const { params } = useRoute();
const { data: restaurant, isError, error } = useFetchRestaurant({ restaurantId: params.restaurantId });

function getFormattedPhone(phone: string) {
  return phone.replace(/(\d{3})(\d{3})(\d{4})/, `($1) $2-$3`);
}
</script>

<template>
  <VAlert v-if="isError" type="warning" class="mt-4" theme="light" text="warning" density="default" variant="tonal">
    {{ (error as unknown as Error).message ?? null }}
    <br>
  </VAlert>
  <div v-else class="restaurant-container grid grid-cols-[minmax(0,_1fr)_16rem] gap-6">
    <VCard v-if="restaurant">
      <VImg
        v-for="photo in restaurant.photos"
        :key="photo"
        :src="restaurant.photos[0]"
        height="250"
        cover
        gradient="to top, rgba(0,0,0,.1), rgba(0,0,0,.5)"
      >
        <VCardTitle class="!text-4xl text-white">
          {{ restaurant.name }}
        </VCardTitle>
        <VRating
          hover
          :length="5"
          :size="32"
          :model-value="getRestaurantRating(restaurant.reviews)"
          active-color="#ffd055"
          color="#FFFF"
        />
      </VImg>
      <VCardText>
        <div class="grid grid-cols-2 gap-4">
          <RestaurantLocation :location="restaurant.location" />
          <KeyValue icon="mdi-phone">
            <p class="text-body-1">
              {{ getFormattedPhone(restaurant.phone) }}
            </p>
          </KeyValue>
        </div>
      </VCardText>
    </VCard>
    <aside v-if="restaurant">
      <ul class="pa-0">
        <RestaurantReview :reviews="restaurant.reviews" />
      </ul>
    </aside>
  </div>
</template>

<style lang="css" scoped>
.restaurant-container {
  display: grid;
  grid-template-columns: minmax(0, 1fr) 16rem;
}

@media screen and (max-width: 640px) {
  .restaurant-container {
    display: flex;
    flex-direction: column;
  }
}
</style>
