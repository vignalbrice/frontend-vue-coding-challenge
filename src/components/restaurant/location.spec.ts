import { mount } from '@vue/test-utils';

import RestaurantLocation from './location.vue';
import { type Location } from '~/composables/restaurants';
import restaurantsResponse from '@/mock/restaurants.json';

const location = restaurantsResponse[0].location as Location;

describe(`restaurant-card`, () => {
  test(`no props`, (context) => {
    const wrapper = mount(RestaurantLocation, {
      global: { plugins: [context.router] },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it(`renders correctly a location`, (context) => {
    const wrapper = mount(RestaurantLocation, {
      global: { plugins: [context.router] },
      props: { location },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
