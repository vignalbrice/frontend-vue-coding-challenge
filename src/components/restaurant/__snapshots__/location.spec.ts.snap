// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`restaurant-card > no props 1`] = `
"<div class=\\"flex gap-2\\"><i class=\\"mdi-map-marker mdi v-icon notranslate v-theme--light v-icon--size-default text-slate-300\\" aria-hidden=\\"true\\"></i>
  <div class=\\"text-slate-600\\"><pre class=\\"my-0\\"></pre>
  </div>
</div>"
`;

exports[`restaurant-card > renders correctly a location 1`] = `
"<div class=\\"flex gap-2\\"><i class=\\"mdi-map-marker mdi v-icon notranslate v-theme--light v-icon--size-default text-slate-300\\" aria-hidden=\\"true\\"></i>
  <div class=\\"text-slate-600\\"><pre class=\\"my-0\\">Jungfernstieg 16 - 20
4. OG
20354 Hamburg
Germany</pre>
  </div>
</div>"
`;
